### DISCUSSION TITLE
Wrapping a textcat in a custom component

### DISCUSSION THREAD

**Discussion by NixBiks** [explosion/spaCy#10182] [Help: Coding & Implementations]
I have a custom component that runs text categorization on each sentence in a `Doc`. This was build in a kinda hacky way in `v2`. Now I want to update it to work in `v3` with the new config system.

Initially I was thinking something like this 

```python
from spacy.pipeline.textcat_multilabel import MultiLabel_TextCategorizer


@Language.factory("my_component", default_config={"textcat": None})
def create_my_component(nlp, name, textcat: MultiLabel_TextCategorizer):
    return MyComponent(textcat)


class MyComponent:
    def __init__(self, textcat):
        this.textcat = textcat
```

and with a config like this

```cfg
[components.my_component]
factory = "my_component"

[components.my_component.textcat]
factory = "textcat_multilabel"
scorer = {"@scorers":"spacy.textcat_multilabel_scorer.v1"}
threshold = 0.5

[components.my_component.textcat.model]
@architectures = "spacy.TextCatBOW.v2"
exclusive_classes = false
ngram_size = 1
no_output_layer = false
nO = null
```

I have my component registered fine but I'm getting the following error when I'm using `spacy.load`

```
Config validation error

my_component -> textcat   instance of MultiLabel_TextCategorizer expected

{
  'nlp': <spacy.lang.en.Englishobjectat0x117057f10>,
  'name': 'my_component',
  'textcat': {
    'factory': 'textcat_multilabel',
    'scorer': {
      '@scorers': 'spacy.textcat_multilabel_scorer.v1'
    },
    'threshold': 0.5,
    'model': {
      '@architectures': 'spacy.TextCatBOW.v2',
      'exclusive_classes': False,
      'ngram_size': 1,
      'no_output_layer': False,
      'nO': None
    }
  },
  '@factories': 'my_component'
}
```

I didn't expect it to work since I still need to point to where my `textcat` model can actually be found but yeah...

**Comment by darrkj** [comment]
Is the issue in having the signatures not aligned. The factory passes`(nlp, name, textcat: MultiLabel_TextCategorizer)` to `create_my_component` but in the __init__ method you only supply `self, textcat` so where it expects to get a nlp it gets the textcat component. That json shows the three things the factory should supply but the error is stating that __init__ is expecting an instance of MultiLabel_TextCategorizer.

**Reply by NixBiks** [reply]
Ah yes that was actually just a typo in this issue. I do have `textcat` in my actual code. I've updated my question accordingly

**Comment by NixBiks** [comment]
#### Workaround

My current workaround is to have only this part in my config

```
[components.my_component]
factory = "my_component"
```

Then I load the component like this

```python
import spacy

from pathlib import Path
from spacy.pipeline.textcat_multilabel import MultiLabel_TextCategorizer


@Language.factory("my_component", default_config={"textcat": None})
def create_my_component(nlp, name, textcat: MultiLabel_TextCategorizer):
    textcat = spacy.load(
        Path(__file__).parent / "model", vocab=nlp.vocab
    ).get_pipe("textcat_multilabel")
    return MyComponent(textcat)


class MyComponent:
    def __init__(self, textcat):
        this.textcat = textcat
```

Note this requires an entire pipeline in `./model` folder with it's own `config.cfg`. So I don't see this as optimal really

**Comment by polm** [comment]
Sorry for the delayed reply to this. 

Have you looked at the [spancat](https://spacy.io/api/spancategorizer)? It's still not as well documented as our other components, but it's a pretty natural fit for labelling each sentence in a document. 

About your component / issue here more generally, it might be easier to subclass the textcat and modify some of the methods on it to get the behavior your want rather than putting it inside another component. Things like serialization etc. should also just work.

Another approach is of course to split your docs up, using one doc for each sentence, and then using a normal pipeline with textcat and reassembling the docs later.
